beacon-chain/rpc/eth/rewards/handlers.go

```diff
@@ -166,22 +178,318 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
+			Code:    http.StatusInternalServerError,
+		}
+		network.WriteError(w, errJson)
+		return
+	}
+
+	totalRewards := make([]AttReward, len(valIndices))
+	for i, v := range valIndices {
+		totalRewards[i] = &TotalAttestationReward{ValidatorIndex: strconv.FormatUint(uint64(v), 10)}
+	}
+	err = attestationsDelta(totalRewards, st, bal, filteredVals)
+	if err != nil {
+		errJson := &network.DefaultErrorJson{
+			Message: "Could not get attestations delta: " + err.Error(),
+			Code:    http.StatusInternalServerError,
+		}
+		network.WriteError(w, errJson)
+		return
+	}
+
+	ir := make([]IdealAttestationReward, len(idealRewards))
+	for i, r := range idealRewards {
+		ir[i] = *r.(*IdealAttestationReward)
+	}
+	tr := make([]TotalAttestationReward, len(totalRewards))
+	for i, r := range totalRewards {
+		tr[i] = *r.(*TotalAttestationReward)
+	}
+	resp := &AttestationRewardsResponse{
+		Data: AttestationRewards{
+			IdealRewards: ir,
+			TotalRewards: tr,
+		},
+	}
+
+	network.WriteJson(w, resp)
+}
+
+func attestationsDelta(
+	rewards []AttReward,
+	beaconState state.BeaconState,
+	bal *precompute.Balance,
+	vals []*precompute.Validator,
+) error {
+	cfg := params.BeaconConfig()
+	prevEpoch := time.PrevEpoch(beaconState)
+	finalizedEpoch := beaconState.FinalizedCheckpointEpoch()
+	increment := cfg.EffectiveBalanceIncrement
+	factor := cfg.BaseRewardFactor
+	baseRewardMultiplier := increment * factor / math.CachedSquareRoot(bal.ActiveCurrentEpoch)
+	leak := helpers.IsInInactivityLeak(prevEpoch, finalizedEpoch)
+
+	// Modified in Altair and Bellatrix.
+	bias := cfg.InactivityScoreBias
+	inactivityPenaltyQuotient, err := beaconState.InactivityPenaltyQuotient()
+	if err != nil {
+		return err
+	}
+	inactivityDenominator := bias * inactivityPenaltyQuotient
+
+	for i, r := range rewards {
+		err = attestationDelta(r, bal, vals[i], baseRewardMultiplier, inactivityDenominator, leak)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func attestationDelta(
+	reward AttReward,
+	bal *precompute.Balance,
+	val *precompute.Validator,
+	baseRewardMultiplier, inactivityDenominator uint64,
+	inactivityLeak bool) error {
+	// TODO: Move this outside
+	eligible := val.IsActivePrevEpoch || (val.IsSlashed && !val.IsWithdrawableCurrentEpoch)
+	// Per spec `ActiveCurrentEpoch` can't be 0 to process attestation delta.
+	if !eligible || bal.ActiveCurrentEpoch == 0 {
+		return nil
+	}
+
+	cfg := params.BeaconConfig()
+	increment := cfg.EffectiveBalanceIncrement
+	effectiveBalance := val.CurrentEpochEffectiveBalance
+	baseReward := (effectiveBalance / increment) * baseRewardMultiplier
+	activeIncrement := bal.ActiveCurrentEpoch / increment
+
+	weightDenominator := cfg.WeightDenominator
+	srcWeight := cfg.TimelySourceWeight
+	tgtWeight := cfg.TimelyTargetWeight
+	headWeight := cfg.TimelyHeadWeight
+	// Process source reward / penalty
+	if val.IsPrevEpochSourceAttester && !val.IsSlashed {
+		if inactivityLeak {
```

Comment:
The different logic here and in the corresponding `attestationDelta` function in the `precompute` package makes this hard to review and is bug prone.